manim/sphinxext/manim_directive.py

```diff
@@ -0,0 +1,83 @@
+
+from docutils.parsers.rst import directives, Directive
+from docutils.parsers.rst.directives.images import Image
+
+import jinja2
+import os
+from os.path import relpath
+
+import shutil
+
+
+class ManimDirective(Directive):
+    r"""Implementation of a ``.. manim::`` directive.
+
+    """
+    has_content = True
+    required_arguments = 1
+    optional_arguments = 1
+    final_argument_whitespace = False
+
+    def run(self):
+        print(self.arguments)
+        clsname = self.arguments[0]
+        include_source = True
+        if len(self.arguments) == 2:
+            include_source = self.arguments[1] == "True"
+
+        state_machine = self.state_machine
+        document = state_machine.document
+
+        source_file_name = document.attributes['source']
+        source_rel_name = relpath(source_file_name, setup.confdir)
+        source_rel_dir = os.path.dirname(source_rel_name)
+        while source_rel_dir.startswith(os.path.sep):
+            source_rel_dir = source_rel_dir[1:]
+
+        dest_dir = os.path.abspath(os.path.join(setup.app.builder.outdir,
+                                            source_rel_dir))
+        print(dest_dir)
+        if not os.path.exists(dest_dir):
+            os.makedirs(dest_dir)
+
+        source_block = ['.. code-block:: python', '', 
+                        *['    ' + line for line in self.content]]
+        source_block = '\n'.join(source_block)
+
+        rendered_template = jinja2.Template(TEMPLATE).render(
+            include_source=include_source,
+            source_block=source_block,
+            clsname=clsname
+        )
+        state_machine.insert_input(rendered_template.split('\n'), 
+                                   source=document.attributes['source'])
+        exec('\n'.join(self.content) + f'\n\n{clsname}()', globals())
```

Comment:
Question: I'm a bit confused as to when in the process manim is actually being executed. I think it's this line, no? Could you expand upon this a bit? It seems to be directly executing what's in the content of the code block. Does this mean that every scene example will have to include `from manim import *` at the top, for example?

Reply:
Yes, you are right. This line takes the code given to the extension, together with the provided class name, and first executes the block and then calls `clsname()`. This triggers building the documentation.

(Currently, this is a bit (huge understatement) rough, as there are no customization options _at all_. This is the main point I'd like to still improve, before this PR is actually ready.)